// @renproject/chains-bitcoin/module/APIs/blockbook.js
// import https from "https";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { utils } from "@renproject/utils";
import { fixUTXO, fixValue, sortUTXOs } from "./API";
export class Blockbook {
    constructor(url) {
        this.fetchHeight = () => __awaiter(this, void 0, void 0, function* () {
            return (yield utils.GET(`${this.url}`)).bestHeight.toString();
        });
        this.fetchUTXOs = (address) => __awaiter(this, void 0, void 0, function* () {
            const url = `${this.url}/utxo/${address}`;
            const response = yield utils.GET(url);
            const data = typeof response === "string" ? JSON.parse(response) : response;
            return (yield Promise.all(data
                .map((utxo) => ({
                txid: utxo.txid,
                amount: utils.isDefined(utxo.satoshis)
                    ? utxo.satoshis.toString()
                    : utils.isDefined(utxo.amount)
                        ? fixValue(utxo.amount, 8).toFixed()
                        : undefined,
                txindex: utxo.vout.toString(),
                height: utxo.height && utxo.height > 0
                    ? utxo.height.toString()
                    : null,
            }))
                // If the amount is undefined, fetch the UTXO again.
                // This is due to the Digibyte explorer not returning
                // amounts correctly when fetching UTXOs.
                .map((utxo) => utils.isDefined(utxo.amount)
                ? utxo
                : this.fetchUTXO(utxo.txid, utxo.txindex)))).sort(sortUTXOs);
        });
        this.fetchUTXO = (txid, txindex) => __awaiter(this, void 0, void 0, function* () {
            const url = `${this.url}/tx/${txid}`;
            const tx = yield utils.GET(url);
            return fixUTXO({
                txid,
                amount: tx.vout[parseInt(txindex, 10)].value.toString(),
                txindex,
                height: tx.blockheight && tx.blockheight > 0
                    ? tx.blockheight.toString()
                    : null,
            }, 8);
        });
        this.broadcastTransaction = (txHex) => __awaiter(this, void 0, void 0, function* () {
            const url = `${this.url}/tx/send`;
            const response = yield utils.POST(url, { rawtx: txHex });
            if (response.error) {
                throw new Error(response.error);
            }
            return response.txid;
        });
        this.url = url.replace(/\/$/, "");
    }
}
