// rubic-sdk/lib/features/cross-chain/providers/via-trade-provider/via-cross-chain-trade.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViaCrossChainTrade = void 0;
var core_1 = require("../../../../core");
var router_sdk_1 = require("@viaprotocol/router-sdk");
var common_1 = require("../../../../common");
var via_default_api_key_1 = require("./constants/via-default-api-key");
var injector_1 = require("../../../../core/sdk/injector");
var features_1 = require("../../..");
var bignumber_js_1 = __importDefault(require("bignumber.js"));
var contract_data_1 = require("./constants/contract-data");
var common_cross_chain_abi_1 = require("../common/constants/common-cross-chain-abi");
var decode_method_1 = require("../../utils/decode-method");
var erc_20_abi_1 = require("../../../../core/blockchain/constants/erc-20-abi");
var swap_request_error_1 = require("../../../../common/errors/swap/swap-request.error");
var not_whitelisted_provider_error_1 = require("../../../../common/errors/swap/not-whitelisted-provider.error");
var empty_address_1 = require("../../../../core/blockchain/constants/empty-address");
var ViaCrossChainTrade = /** @class */ (function (_super) {
    __extends(ViaCrossChainTrade, _super);
    function ViaCrossChainTrade(crossChainTrade, providerAddress) {
        var _this = _super.call(this, providerAddress) || this;
        _this.type = features_1.CROSS_CHAIN_TRADE_TYPE.VIA;
        _this.via = new router_sdk_1.Via(via_default_api_key_1.VIA_DEFAULT_CONFIG);
        _this.from = crossChainTrade.from;
        _this.to = crossChainTrade.to;
        _this.route = crossChainTrade.route;
        _this.gasData = crossChainTrade.gasData;
        _this.priceImpact = crossChainTrade.priceImpact;
        _this.toTokenAmountMin = crossChainTrade.toTokenAmountMin;
        _this.feeInfo = crossChainTrade.feeInfo;
        _this.cryptoFeeToken = crossChainTrade.cryptoFeeToken;
        _this.itType = crossChainTrade.itType;
        _this.bridgeType = crossChainTrade.bridgeType;
        _this.fromWeb3Public = injector_1.Injector.web3PublicService.getWeb3Public(_this.from.blockchain);
        return _this;
    }
    /** @internal */
    ViaCrossChainTrade.getGasData = function (from, to, route) {
        return __awaiter(this, void 0, void 0, function () {
            var fromBlockchain, walletAddress, _a, contractAddress, contractAbi, methodName, methodArguments, value, web3Public, _b, gasLimit, gasPrice, _c, _d, _e, _f, increasedGasLimit, _err_1;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        fromBlockchain = from.blockchain;
                        walletAddress = injector_1.Injector.web3Private.address;
                        if (!walletAddress) {
                            return [2 /*return*/, null];
                        }
                        _g.label = 1;
                    case 1:
                        _g.trys.push([1, 5, , 6]);
                        return [4 /*yield*/, new ViaCrossChainTrade({
                                from: from,
                                to: to,
                                route: route,
                                gasData: null,
                                priceImpact: 0,
                                toTokenAmountMin: new bignumber_js_1.default(0),
                                feeInfo: {
                                    fixedFee: null,
                                    platformFee: null,
                                    cryptoFee: null
                                },
                                cryptoFeeToken: {},
                                itType: { from: undefined, to: undefined },
                                bridgeType: features_1.BRIDGE_TYPE.DE_BRIDGE
                            }, empty_address_1.EMPTY_ADDRESS).getContractParams({})];
                    case 2:
                        _a = _g.sent(), contractAddress = _a.contractAddress, contractAbi = _a.contractAbi, methodName = _a.methodName, methodArguments = _a.methodArguments, value = _a.value;
                        web3Public = injector_1.Injector.web3PublicService.getWeb3Public(fromBlockchain);
                        _d = (_c = Promise).all;
                        _e = [web3Public.getEstimatedGas(contractAbi, contractAddress, methodName, methodArguments, walletAddress, value)];
                        _f = bignumber_js_1.default.bind;
                        return [4 /*yield*/, injector_1.Injector.gasPriceApi.getGasPrice(from.blockchain)];
                    case 3: return [4 /*yield*/, _d.apply(_c, [_e.concat([
                                new (_f.apply(bignumber_js_1.default, [void 0, _g.sent()]))()
                            ])])];
                    case 4:
                        _b = _g.sent(), gasLimit = _b[0], gasPrice = _b[1];
                        if (!(gasLimit === null || gasLimit === void 0 ? void 0 : gasLimit.isFinite())) {
                            return [2 /*return*/, null];
                        }
                        increasedGasLimit = core_1.Web3Pure.calculateGasMargin(gasLimit, 1.2);
                        return [2 /*return*/, {
                                gasLimit: increasedGasLimit,
                                gasPrice: gasPrice
                            }];
                    case 5:
                        _err_1 = _g.sent();
                        return [2 /*return*/, null];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(ViaCrossChainTrade.prototype, "fromContractAddress", {
        get: function () {
            return contract_data_1.viaContractAddress[this.from.blockchain];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ViaCrossChainTrade.prototype, "uuid", {
        get: function () {
            return this.route.actions[0].uuid;
        },
        enumerable: false,
        configurable: true
    });
    ViaCrossChainTrade.prototype.checkTradeErrors = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.checkWalletConnected();
                        this.checkBlockchainCorrect();
                        return [4 /*yield*/, this.checkUserBalance()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    ViaCrossChainTrade.prototype.swap = function (options) {
        if (options === void 0) { options = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var onConfirm, gasLimit, gasPrice, _a, contractAddress, contractAbi, methodName, methodArguments, value, transactionHash, onTransactionHash, _b, err_1;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.checkTradeErrors()];
                    case 1:
                        _c.sent();
                        return [4 /*yield*/, this.checkAllowanceAndApprove(options)];
                    case 2:
                        _c.sent();
                        features_1.CrossChainTrade.checkReceiverAddress(options === null || options === void 0 ? void 0 : options.receiverAddress);
                        onConfirm = options.onConfirm, gasLimit = options.gasLimit, gasPrice = options.gasPrice;
                        return [4 /*yield*/, this.getContractParams(options)];
                    case 3:
                        _a = _c.sent(), contractAddress = _a.contractAddress, contractAbi = _a.contractAbi, methodName = _a.methodName, methodArguments = _a.methodArguments, value = _a.value;
                        onTransactionHash = function (hash) {
                            if (onConfirm) {
                                onConfirm(hash);
                            }
                            transactionHash = hash;
                        };
                        _c.label = 4;
                    case 4:
                        _c.trys.push([4, 10, , 11]);
                        return [4 /*yield*/, injector_1.Injector.web3Private.tryExecuteContractMethod(contractAddress, contractAbi, methodName, methodArguments, {
                                gas: gasLimit,
                                gasPrice: gasPrice,
                                value: value,
                                onTransactionHash: onTransactionHash
                            })];
                    case 5:
                        _c.sent();
                        _c.label = 6;
                    case 6:
                        _c.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, this.via.startRoute({
                                fromAddress: this.fromContractAddress,
                                toAddress: (options === null || options === void 0 ? void 0 : options.receiverAddress) || this.walletAddress,
                                routeId: this.route.routeId,
                                txHash: transactionHash
                            })];
                    case 7:
                        _c.sent();
                        return [3 /*break*/, 9];
                    case 8:
                        _b = _c.sent();
                        return [3 /*break*/, 9];
                    case 9: return [2 /*return*/, transactionHash];
                    case 10:
                        err_1 = _c.sent();
                        if (err_1 instanceof common_1.FailedToCheckForTransactionReceiptError) {
                            return [2 /*return*/, transactionHash];
                        }
                        if ([400, 500, 503].includes(err_1.code)) {
                            throw new swap_request_error_1.SwapRequestError();
                        }
                        throw err_1;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    ViaCrossChainTrade.prototype.getContractParams = function (options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var swapTransaction, toChainId, providerRouter, swapArguments, methodArguments, providerGateway, approveTransaction, decodedData, sourceValue, fixedFee, value;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.via.buildTx({
                            routeId: this.route.routeId,
                            fromAddress: this.fromContractAddress,
                            receiveAddress: (options === null || options === void 0 ? void 0 : options.receiverAddress) || this.walletAddress,
                            numAction: 0
                        })];
                    case 1:
                        swapTransaction = _c.sent();
                        toChainId = core_1.BlockchainsInfo.getBlockchainByName(this.to.blockchain).id;
                        providerRouter = swapTransaction.to;
                        swapArguments = [
                            this.from.address,
                            this.from.stringWeiAmount,
                            toChainId,
                            this.to.address,
                            core_1.Web3Pure.toWei(this.toTokenAmountMin, this.to.decimals),
                            (options === null || options === void 0 ? void 0 : options.receiverAddress) || this.walletAddress,
                            this.providerAddress,
                            providerRouter
                        ];
                        methodArguments = [swapArguments];
                        if (!!this.from.isNative) return [3 /*break*/, 3];
                        return [4 /*yield*/, this.via.buildApprovalTx({
                                owner: this.fromContractAddress,
                                routeId: this.route.routeId,
                                numAction: 0
                            })];
                    case 2:
                        approveTransaction = _c.sent();
                        decodedData = decode_method_1.MethodDecoder.decodeMethod(erc_20_abi_1.ERC20_TOKEN_ABI.find(function (method) { return method.name === 'approve'; }), approveTransaction.data);
                        providerGateway = decodedData.params.find(function (param) { return param.name === '_spender'; }).value;
                        methodArguments.push(providerGateway);
                        _c.label = 3;
                    case 3:
                        methodArguments.push(swapTransaction.data);
                        return [4 /*yield*/, this.checkProviderIsWhitelisted(providerRouter, providerGateway)];
                    case 4:
                        _c.sent();
                        sourceValue = this.from.isNative ? this.from.stringWeiAmount : '0';
                        fixedFee = core_1.Web3Pure.toWei(((_b = (_a = this.feeInfo) === null || _a === void 0 ? void 0 : _a.fixedFee) === null || _b === void 0 ? void 0 : _b.amount) || 0);
                        value = new bignumber_js_1.default(sourceValue).plus(fixedFee).toFixed(0);
                        return [2 /*return*/, {
                                contractAddress: this.fromContractAddress,
                                contractAbi: common_cross_chain_abi_1.commonCrossChainAbi,
                                methodName: this.methodName,
                                methodArguments: methodArguments,
                                value: value
                            }];
                }
            });
        });
    };
    ViaCrossChainTrade.prototype.checkProviderIsWhitelisted = function (providerRouter, providerGateway) {
        return __awaiter(this, void 0, void 0, function () {
            var whitelistedContracts;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, injector_1.Injector.web3PublicService
                            .getWeb3Public(this.from.blockchain)
                            .callContractMethod(this.fromContractAddress, contract_data_1.viaContractAbi, 'getAvailableRouters')];
                    case 1:
                        whitelistedContracts = _a.sent();
                        if (!whitelistedContracts.find(function (whitelistedContract) {
                            return (0, common_1.compareAddresses)(whitelistedContract, providerRouter);
                        }) ||
                            (providerGateway &&
                                !whitelistedContracts.find(function (whitelistedContract) {
                                    return (0, common_1.compareAddresses)(whitelistedContract, providerGateway);
                                }))) {
                            throw new not_whitelisted_provider_error_1.NotWhitelistedProviderError(providerRouter, providerGateway);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    ViaCrossChainTrade.prototype.getTradeAmountRatio = function (fromUsd) {
        var usdCryptoFee = this.cryptoFeeToken.price.multipliedBy(this.cryptoFeeToken.tokenAmount);
        return fromUsd.plus(usdCryptoFee).dividedBy(this.to.tokenAmount);
    };
    return ViaCrossChainTrade;
}(features_1.CrossChainTrade));
exports.ViaCrossChainTrade = ViaCrossChainTrade;
