// @renproject/provider/methods/ren_submitTx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.crossChainParamsType = exports.burnParamsType = void 0;
const utils_1 = require("@renproject/utils");
exports.burnParamsType = {
    struct: [
        {
            amount: utils_1.PackPrimitive.U256,
        },
        {
            to: utils_1.PackPrimitive.Str,
        },
        {
            nonce: utils_1.PackPrimitive.Bytes32,
        },
    ],
};
exports.crossChainParamsType = {
    struct: [
        {
            txid: utils_1.PackPrimitive.Bytes,
        },
        {
            txindex: utils_1.PackPrimitive.U32,
        },
        {
            amount: utils_1.PackPrimitive.U256,
        },
        {
            payload: utils_1.PackPrimitive.Bytes,
        },
        {
            phash: utils_1.PackPrimitive.Bytes32,
        },
        {
            to: utils_1.PackPrimitive.Str,
        },
        {
            nonce: utils_1.PackPrimitive.Bytes32,
        },
        {
            nhash: utils_1.PackPrimitive.Bytes32,
        },
        {
            gpubkey: utils_1.PackPrimitive.Bytes,
        },
        {
            ghash: utils_1.PackPrimitive.Bytes32,
        },
    ],
};
