// rubic-sdk/lib/core/blockchain/constants/blockchains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockchains = void 0;
var native_token_address_1 = require("./native-token-address");
var blockchain_name_1 = require("../models/blockchain-name");
var token_1 = require("../tokens/token");
exports.blockchains = [
    {
        id: 1,
        name: blockchain_name_1.BLOCKCHAIN_NAME.ETHEREUM,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.ETHEREUM,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'Ethereum',
            symbol: 'ETH',
            decimals: 18
        })
    },
    {
        id: 56,
        name: blockchain_name_1.BLOCKCHAIN_NAME.BINANCE_SMART_CHAIN,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.BINANCE_SMART_CHAIN,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'Binance Coin',
            symbol: 'BNB',
            decimals: 18
        })
    },
    {
        id: 137,
        name: blockchain_name_1.BLOCKCHAIN_NAME.POLYGON,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.POLYGON,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'Matic Network',
            symbol: 'MATIC',
            decimals: 18
        })
    },
    {
        id: 43114,
        name: blockchain_name_1.BLOCKCHAIN_NAME.AVALANCHE,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.AVALANCHE,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'AVAX',
            symbol: 'AVAX',
            decimals: 18
        })
    },
    {
        id: 1285,
        name: blockchain_name_1.BLOCKCHAIN_NAME.MOONRIVER,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.MOONRIVER,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'MOVR',
            symbol: 'MOVR',
            decimals: 18
        })
    },
    {
        id: 250,
        name: blockchain_name_1.BLOCKCHAIN_NAME.FANTOM,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.FANTOM,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'Fantom',
            symbol: 'FTM',
            decimals: 18
        })
    },
    {
        id: 1666600000,
        name: blockchain_name_1.BLOCKCHAIN_NAME.HARMONY,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.HARMONY,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'ONE',
            symbol: 'ONE',
            decimals: 18
        })
    },
    {
        id: 42161,
        name: blockchain_name_1.BLOCKCHAIN_NAME.ARBITRUM,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.ARBITRUM,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'AETH',
            symbol: 'AETH',
            decimals: 18
        })
    },
    {
        id: 1313161554,
        name: blockchain_name_1.BLOCKCHAIN_NAME.AURORA,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.AURORA,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'aETH',
            symbol: 'aETH',
            decimals: 18
        })
    },
    {
        id: 40,
        name: blockchain_name_1.BLOCKCHAIN_NAME.TELOS,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.TELOS,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'Telos EVM',
            symbol: 'TLOS',
            decimals: 18
        })
    },
    {
        id: 10,
        name: blockchain_name_1.BLOCKCHAIN_NAME.OPTIMISM,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.OPTIMISM,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'ETH',
            symbol: 'ETH',
            decimals: 18
        })
    },
    {
        id: 25,
        name: blockchain_name_1.BLOCKCHAIN_NAME.CRONOS,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.CRONOS,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'CRO',
            symbol: 'CRO',
            decimals: 18
        })
    },
    {
        id: 66,
        name: blockchain_name_1.BLOCKCHAIN_NAME.OKE_X_CHAIN,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.OKE_X_CHAIN,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'OKT',
            symbol: 'OKT',
            decimals: 18
        })
    },
    {
        id: 100,
        name: blockchain_name_1.BLOCKCHAIN_NAME.GNOSIS,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.GNOSIS,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'xDAI',
            symbol: 'xDAI',
            decimals: 18
        })
    },
    {
        id: 122,
        name: blockchain_name_1.BLOCKCHAIN_NAME.FUSE,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.FUSE,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'FUSE',
            symbol: 'FUSE',
            decimals: 18
        })
    },
    {
        id: 1284,
        name: blockchain_name_1.BLOCKCHAIN_NAME.MOONBEAM,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.MOONBEAM,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'GLMR',
            symbol: 'GLMR',
            decimals: 18
        })
    },
    {
        id: 42220,
        name: blockchain_name_1.BLOCKCHAIN_NAME.CELO,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.CELO,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'CELO',
            symbol: 'CELO',
            decimals: 18
        })
    },
    {
        id: 288,
        name: blockchain_name_1.BLOCKCHAIN_NAME.BOBA,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.BOBA,
            address: native_token_address_1.NATIVE_TOKEN_ADDRESS,
            name: 'ETH',
            symbol: 'ETH',
            decimals: 18
        })
    },
    {
        id: 5555,
        name: blockchain_name_1.BLOCKCHAIN_NAME.BITCOIN,
        nativeCoin: new token_1.Token({
            blockchain: blockchain_name_1.BLOCKCHAIN_NAME.BITCOIN,
            address: '',
            name: 'Bitcoin',
            symbol: 'BTC',
            decimals: 8
        })
    }
];
