// rubic-sdk/lib/features/cross-chain/providers/via-trade-provider/constants/contract-data.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.viaContractAbi = exports.viaContractAddress = void 0;
var common_cross_chain_abi_1 = require("../../common/constants/common-cross-chain-abi");
var via_cross_chain_supported_blockchain_1 = require("./via-cross-chain-supported-blockchain");
var rubic_proxy_contract_address_1 = require("../../../constants/rubic-proxy-contract-address");
exports.viaContractAddress = via_cross_chain_supported_blockchain_1.viaCrossChainSupportedBlockchains.reduce(function (acc, blockchain) {
    var _a;
    return __assign(__assign({}, acc), (_a = {}, _a[blockchain] = rubic_proxy_contract_address_1.rubicProxyContractAddress[blockchain], _a));
}, {});
exports.viaContractAbi = common_cross_chain_abi_1.commonCrossChainAbi.concat({
    inputs: [],
    name: 'getAvailableRouters',
    outputs: [{ internalType: 'address[]', name: '', type: 'address[]' }],
    stateMutability: 'view',
    type: 'function'
});
