// @renproject/utils/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorWithCode = exports.RenJSError = void 0;
const extractError_1 = require("./internal/extractError");
/**
 * RenJS error codes - errors thrown from RenJS are in the process of being
 * converted to `ErrorWithCode` errors with the following error codes.
 */
var RenJSError;
(function (RenJSError) {
    // General errors.
    RenJSError["UNKNOWN_ERROR"] = "RENJS:UNKNOWN_ERROR";
    RenJSError["NOT_IMPLEMENTED"] = "RENJS:NOT_IMPLEMENTED";
    RenJSError["PARAMETER_ERROR"] = "RENJS:PARAMETER_ERROR";
    RenJSError["INTERNAL_ERROR"] = "RENJS:INTERNAL_ERROR";
    RenJSError["NETWORK_ERROR"] = "RENJS:NETWORK_ERROR";
    RenJSError["SIGNER_ERROR"] = "RENJS:SIGNER_ERROR";
    RenJSError["TRANSACTION_NOT_FOUND"] = "RENJS:TRANSACTION_NOT_FOUND";
    // Indicates that the RenVM transaction has returned with a Reverted status
    // and a revert reason. A reverted transaction cannot be re-submitted.
    RenJSError["RENVM_TRANSACTION_REVERTED"] = "RENJS:RENVM_TRANSACTION_REVERTED";
    // Indicates that the chain transaction reverted. It may be possible to
    // resubmit the transaction.
    RenJSError["CHAIN_TRANSACTION_REVERTED"] = "RENJS:CHAIN_TRANSACTION_REVERTED";
    // Indicates that submitting the gateway details failed.
    RenJSError["GATEWAY_SUBMISSION_FAILED"] = "RENJS:GATEWAY_SUBMISSION_FAILED";
    RenJSError["INCORRECT_PROVIDER_NETWORK"] = "RENJS:INCORRECT_PROVIDER_NETWORK";
    RenJSError["INCORRECT_SIGNER_NETWORK"] = "RENJS:INCORRECT_SIGNER_NETWORK";
})(RenJSError = exports.RenJSError || (exports.RenJSError = {}));
/**
 * An ErrorWithCode is an Error instance with an additional `code` field.
 * Because this is a common pattern with various implementations provided
 * by libraries, checking if an Error is an ErrorWithCode should be done using
 * `isErrorWithCode(error)` instead of `error instanceof ErrorWithCode`.
 */
class ErrorWithCode extends Error {
    /**
     * @param message An error message, passed on to the Error constructor.
     * @param code An error code, defined in a standard manner to allow for
     * easier error handling.
     */
    constructor(message, code, prefix) {
        super(prefix
            ? `${prefix}: ${(0, extractError_1.extractError)(message)}`
            : (0, extractError_1.extractError)(message));
        this.code = code;
    }
}
exports.ErrorWithCode = ErrorWithCode;
/**
 * Check if the provided value is an Error instance and has a code field which
 * contains a string.
 */
ErrorWithCode.isErrorWithCode = (error) => 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
error instanceof Error && typeof error.code === "string";
/**
 * Add an error code to an existing Error instance.
 */
ErrorWithCode.updateError = (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
error, code, errorPrefix) => {
    // Check the code passed in is a RenJS error. If it isn't, it may have
    // been set by a dependency.
    if (typeof code !== "string" || code.slice(0, 6) !== "RENJS:") {
        code = RenJSError.INTERNAL_ERROR;
    }
    if (error instanceof Error) {
        error.code = code;
        if (errorPrefix) {
            error.message = `${errorPrefix}: ${(0, extractError_1.extractError)(error)}`;
        }
        return error;
    }
    else {
        const message = (0, extractError_1.extractError)(error);
        return new ErrorWithCode(errorPrefix && message
            ? `${errorPrefix}: ${message}}`
            : errorPrefix || message, code);
    }
};
