// @lifi/sdk/dist/execution/bridges/bridge.execute.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { constants } from 'ethers';
import ApiService from '../../services/ApiService';
import ChainsService from '../../services/ChainsService';
import { LifiErrorCode, TransactionError } from '../../utils/errors';
import { getProvider } from '../../utils/getProvider';
import { getTransactionFailedMessage, parseError } from '../../utils/parseError';
import { personalizeStep } from '../../utils/utils';
import { checkAllowance } from '../allowance.execute';
import { balanceCheck } from '../balanceCheck.execute';
import { stepComparison } from '../stepComparison';
import { switchChain } from '../switchChain';
import { getSubstatusMessage, waitForReceivingTransaction } from '../utils';
export class BridgeExecutionManager {
    constructor() {
        this.shouldContinue = true;
        this.setShouldContinue = (val) => {
            this.shouldContinue = val;
        };
        this.execute = ({ signer, step, statusManager, settings, }) => __awaiter(this, void 0, void 0, function* () {
            var _a, _b, _c, _d;
            const { action, estimate } = step;
            step.execution = statusManager.initExecutionObject(step);
            const chainsService = ChainsService.getInstance();
            const fromChain = yield chainsService.getChainById(action.fromChainId);
            const toChain = yield chainsService.getChainById(action.toChainId);
            // STEP 1: Check Allowance ////////////////////////////////////////////////
            // approval still needed?
            const oldCrossProcess = step.execution.process.find((p) => p.type === 'CROSS_CHAIN');
            if (!(oldCrossProcess === null || oldCrossProcess === void 0 ? void 0 : oldCrossProcess.txHash)) {
                if (action.fromToken.address !== constants.AddressZero) {
                    // Check Token Approval only if fromToken is not the native token => no approval needed in that case
                    yield checkAllowance(signer, step, fromChain, action.fromToken, action.fromAmount, estimate.approvalAddress, statusManager, settings.infiniteApproval, this.shouldContinue);
                }
            }
            // STEP 2: Get Transaction ////////////////////////////////////////////////
            let crossChainProcess = statusManager.findOrCreateProcess('CROSS_CHAIN', step);
            try {
                let tx;
                if (crossChainProcess.txHash) {
                    // load exiting transaction
                    tx = yield getProvider(signer).getTransaction(crossChainProcess.txHash);
                }
                else {
                    // check balance
                    yield balanceCheck(signer, step);
                    // create new transaction
                    const personalizedStep = yield personalizeStep(signer, step);
                    const updatedStep = yield ApiService.getStepTransaction(personalizedStep);
                    step = Object.assign(Object.assign({}, (yield stepComparison(statusManager, personalizedStep, updatedStep, settings.acceptSlippageUpdateHook, this.shouldContinue))), { execution: step.execution });
                    const { transactionRequest } = step;
                    if (!transactionRequest) {
                        throw new TransactionError(LifiErrorCode.TransactionUnprepared, 'Unable to prepare transaction.');
                    }
                    // STEP 3: Send Transaction ///////////////////////////////////////////////
                    // make sure that chain is still correct
                    const updatedSigner = yield switchChain(signer, statusManager, step, settings.switchChainHook, this.shouldContinue);
                    if (!updatedSigner) {
                        // chain switch was not successful, stop execution here
                        return step.execution;
                    }
                    signer = updatedSigner;
                    crossChainProcess = statusManager.updateProcess(step, crossChainProcess.type, 'ACTION_REQUIRED');
                    if (!this.shouldContinue) {
                        return step.execution;
                    }
                    tx = yield signer.sendTransaction(transactionRequest);
                    // STEP 4: Wait for Transaction ///////////////////////////////////////////
                    crossChainProcess = statusManager.updateProcess(step, crossChainProcess.type, 'PENDING', {
                        txHash: tx.hash,
                        txLink: fromChain.metamask.blockExplorerUrls[0] + 'tx/' + tx.hash,
                    });
                }
                yield tx.wait();
            }
            catch (e) {
                if (e.code === 'TRANSACTION_REPLACED' && e.replacement) {
                    crossChainProcess = statusManager.updateProcess(step, crossChainProcess.type, 'PENDING', {
                        txHash: e.replacement.hash,
                        txLink: fromChain.metamask.blockExplorerUrls[0] +
                            'tx/' +
                            e.replacement.hash,
                    });
                }
                else {
                    const error = yield parseError(e, step, crossChainProcess);
                    crossChainProcess = statusManager.updateProcess(step, crossChainProcess.type, 'FAILED', {
                        error: {
                            message: error.message,
                            htmlMessage: error.htmlMessage,
                            code: error.code,
                        },
                    });
                    statusManager.updateExecution(step, 'FAILED');
                    throw error;
                }
            }
            crossChainProcess = statusManager.updateProcess(step, crossChainProcess.type, 'DONE');
            // STEP 5: Wait for Receiver //////////////////////////////////////
            let receivingChainProcess = statusManager.findOrCreateProcess('RECEIVING_CHAIN', step, 'PENDING');
            let statusResponse;
            try {
                if (!crossChainProcess.txHash) {
                    throw new Error('Transaction hash is undefined.');
                }
                statusResponse = yield waitForReceivingTransaction(crossChainProcess.txHash, statusManager, receivingChainProcess.type, step);
            }
            catch (e) {
                receivingChainProcess = statusManager.updateProcess(step, receivingChainProcess.type, 'FAILED', {
                    error: {
                        code: LifiErrorCode.TransactionFailed,
                        message: 'Failed while waiting for receiving chain.',
                        htmlMessage: getTransactionFailedMessage(step, crossChainProcess.txLink),
                    },
                });
                statusManager.updateExecution(step, 'FAILED');
                throw e;
            }
            receivingChainProcess = statusManager.updateProcess(step, receivingChainProcess.type, 'DONE', {
                substatus: statusResponse.substatus,
                substatusMessage: statusResponse.substatusMessage ||
                    getSubstatusMessage(statusResponse.status, statusResponse.substatus),
                txHash: (_a = statusResponse.receiving) === null || _a === void 0 ? void 0 : _a.txHash,
                txLink: toChain.metamask.blockExplorerUrls[0] +
                    'tx/' +
                    ((_b = statusResponse.receiving) === null || _b === void 0 ? void 0 : _b.txHash),
            });
            statusManager.updateExecution(step, 'DONE', {
                fromAmount: statusResponse.sending.amount,
                toAmount: (_c = statusResponse.receiving) === null || _c === void 0 ? void 0 : _c.amount,
                toToken: (_d = statusResponse.receiving) === null || _d === void 0 ? void 0 : _d.token,
                gasUsed: statusResponse.sending.gasUsed,
                gasPrice: statusResponse.sending.gasPrice,
            });
            // DONE
            return step.execution;
        });
    }
}
