// @renproject/chains-bitcoin/APIs/API.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CombinedAPI = exports.fixUTXOs = exports.fixUTXO = exports.fixValue = exports.sortUTXOs = exports.DEFAULT_TIMEOUT = void 0;
const utils_1 = require("@renproject/utils");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
// Default timeout for network requests.
const SECONDS = 1000;
exports.DEFAULT_TIMEOUT = 30 * SECONDS;
/**
 * sortUTXOs compares two UTXOs by height, then amount, then txid, then txindex.
 *
 * @returns a negative value to represent that a should come before b or a
 * positive value to represent that b should come before a, or 0 if a and b are
 * equal.
 */
const sortUTXOs = (a, b) => {
    const aHeight = a.height ? new bignumber_js_1.default(a.height) : null;
    const bHeight = b.height ? new bignumber_js_1.default(b.height) : null;
    // Compare heights first
    if (aHeight) {
        if (bHeight) {
            if (!aHeight.isEqualTo(bHeight))
                return bHeight.minus(aHeight).toNumber();
        }
        else {
            return -1;
        }
    }
    else if (bHeight) {
        return 1;
    }
    // If the heights are the same (same number of both null), compare amounts.
    const aAmount = new bignumber_js_1.default(a.amount);
    const bAmount = new bignumber_js_1.default(b.amount);
    if (aAmount.isEqualTo(bAmount)) {
        return bAmount.minus(aAmount).toNumber();
    }
    // If the heights and amounts are equal, compare txid.
    if (a.txid !== b.txid) {
        return a.txid <= b.txid ? -1 : 1;
    }
    // Fall back to the txindex. If the txid and txindex are the same for both
    // transactions, then the two transactions represent the same transfer, and
    // the returned value is `0`.
    const aTxindex = new bignumber_js_1.default(a.txindex);
    const bTxindex = new bignumber_js_1.default(b.txindex);
    return bTxindex.minus(aTxindex).toNumber();
};
exports.sortUTXOs = sortUTXOs;
/**
 * fixValue turns a readable value, e.g. `0.0001` BTC, to the value in the smallest
 * unit, e.g. `10000` sats.
 *
 * @example
 * fixValue(0.0001, 8) = 10000;
 * @param value Value in the readable representation, e.g. `0.0001` BTC.
 * @param decimals The number of decimals to shift by, e.g. 8.
 */
const fixValue = (value, decimals) => new bignumber_js_1.default(value).shiftedBy(decimals).decimalPlaces(0);
exports.fixValue = fixValue;
/**
 * fixUTXO calls {{fixValue}} on the value of the UTXO.
 */
const fixUTXO = (tx, decimals) => (Object.assign(Object.assign({}, tx), { amount: (0, exports.fixValue)(tx.amount, decimals).toFixed() }));
exports.fixUTXO = fixUTXO;
/**
 * fixUTXOs maps over an array of UTXOs and calls {{fixValue}}.
 */
const fixUTXOs = (utxos, decimals) => utxos.map((utxo) => (0, exports.fixUTXO)(utxo, decimals));
exports.fixUTXOs = fixUTXOs;
const notNull = (x) => {
    if (x === undefined || x === null) {
        throw new Error(`Unexpected ${String(x)} value.`);
    }
    return x;
};
const withPriority = (api, defaultPriority = 0) => api.api &&
    api.priority !== undefined
    ? api
    : { api: api, priority: defaultPriority };
class CombinedAPI {
    constructor(apis = [], { priority = 0 } = {}) {
        /**
         * Provide a new API to be used with the other APIs.
         *
         * @param api The API to add.
         * @param config Config for the API, including the priority.
         * @param config.priority Optionally set the priority of the API, where
         * a lower priority means it will be selected before other APIs.
         */
        this.withAPI = (api, { priority = 0 } = {}) => {
            this.apis.push(withPriority(api, priority));
            return this;
        };
        this.fetchHeight = async () => {
            return this.forEachAPI(
            // Filter APIs with `fetchHeight`.
            (api) => api.fetchHeight !== undefined, 
            // Call `fetchHeight` on the API.
            (api) => notNull(api.fetchHeight).bind(api)());
        };
        this.fetchUTXO = async (txid, txindex) => {
            return this.forEachAPI(
            // Filter APIs with `fetchUTXO`.
            (api) => api.fetchUTXO !== undefined, 
            // Call `fetchUTXO` on the API.
            (api) => notNull(api.fetchUTXO).bind(api)(txid, txindex));
        };
        this.fetchUTXOs = async (address, confirmations) => this.forEachAPI(
        // Filter APIs with `fetchUTXOs`.
        (api) => api.fetchUTXOs !== undefined, 
        // Call `fetchUTXOs` on the API.
        (api) => notNull(api.fetchUTXOs).bind(api)(address, confirmations));
        this.fetchTXs = async (address, confirmations) => this.forEachAPI(
        // Filter APIs with `fetchTXs`.
        (api) => api.fetchTXs !== undefined, 
        // Call `fetchTXs` on the API.
        (api) => notNull(api.fetchTXs).bind(api)(address, confirmations));
        this.broadcastTransaction = async (hex) => {
            return this.forEachAPI(
            // Filter APIs with `broadcastTransaction`.
            (api) => api.broadcastTransaction !== undefined, 
            // Call `broadcastTransaction` on the API.
            (api) => notNull(api.broadcastTransaction).bind(api)(hex));
        };
        this.forEachAPI = async (filter, onAPI) => {
            const apis = this.apis
                .map((api, index) => ({ api, index }))
                .filter(({ api }) => filter(api.api))
                .sort(({ api: a }, { api: b }) => 
            // Sort by priority, and randomly for the same priority.
            a.priority !== b.priority
                ? a.priority - b.priority
                : Math.random() * 2 - 1);
            if (!apis.length) {
                throw new Error(`No API available for call.`);
            }
            let firstError;
            const previousIndices = [];
            for (const { api, index } of apis) {
                try {
                    const result = await onAPI(api.api);
                    // If any previous API failed, it may be down or rate limited,
                    // so its priority is reduced.
                    for (const previousIndex of previousIndices) {
                        this.apis[previousIndex].priority -= 5;
                    }
                    return result;
                }
                catch (error) {
                    previousIndices.push(index);
                    firstError =
                        firstError ||
                            (error instanceof Error
                                ? error
                                : new Error(utils_1.utils.extractError(error)));
                }
            }
            throw firstError;
        };
        this.apis = apis.map((api) => withPriority(api, priority));
    }
}
exports.CombinedAPI = CombinedAPI;
