// rubic-sdk/lib/features/cross-chain/cross-chain-status-manager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossChainStatusManager = void 0;
var core_1 = require("../../core");
var injector_1 = require("../../core/sdk/injector");
var celer_cross_chain_event_statuses_abi_1 = require("./providers/celer-trade-provider/constants/celer-cross-chain-event-statuses-abi");
var decode_logs_1 = require("./utils/decode-logs");
var router_sdk_1 = require("@viaprotocol/router-sdk");
var via_default_api_key_1 = require("./providers/via-trade-provider/constants/via-default-api-key");
var via_swap_status_1 = require("./providers/via-trade-provider/models/via-swap-status");
var cross_chain_trade_type_1 = require("./models/cross-chain-trade-type");
var celer_cross_chain_contract_abi_1 = require("./providers/celer-trade-provider/constants/celer-cross-chain-contract-abi");
var celer_cross_chain_contracts_addresses_1 = require("./providers/celer-trade-provider/constants/celer-cross-chain-contracts-addresses");
var celer_swap_status_enum_1 = require("./providers/common/celer-rubic/models/celer-swap-status.enum");
var rubic_cross_chain_contracts_addresses_1 = require("./providers/rubic-trade-provider/constants/rubic-cross-chain-contracts-addresses");
var rubic_swap_status_enum_1 = require("./providers/common/celer-rubic/models/rubic-swap-status.enum");
var processed_transactios_method_abi_1 = require("./providers/common/celer-rubic/constants/processed-transactios-method-abi");
var cross_chain_tx_status_1 = require("./models/cross-chain-tx-status");
var lifi_swap_status_1 = require("./providers/lifi-trade-provider/models/lifi-swap-status");
var symbiosis_swap_status_1 = require("./providers/symbiosis-trade-provider/models/symbiosis-swap-status");
/**
 * Contains methods for getting cross-chain trade statuses.
 */
var CrossChainStatusManager = /** @class */ (function () {
    function CrossChainStatusManager() {
        var _a;
        this.httpClient = injector_1.Injector.httpClient;
        this.getDstTxStatusFnMap = (_a = {},
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.CELER] = this.getCelerDstSwapStatus,
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.RUBIC] = this.getRubicDstSwapStatus,
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.LIFI] = this.getLifiDstSwapStatus,
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.SYMBIOSIS] = this.getSymbiosisDstSwapStatus,
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.DEBRIDGE] = this.getDebridgeDstSwapStatus,
            _a[cross_chain_trade_type_1.CROSS_CHAIN_TRADE_TYPE.VIA] = this.getViaDstSwapStatus,
            _a);
    }
    /**
     * Returns cross-chain trade statuses on the source and target network.
     * The result consists of the status of the source and target transactions.
     * @example
     * ```ts
     * const tradeData = {
     *   fromBlockchain: BLOCKCHAIN_NAME.FANTOM,
     *   toBlockchain: BLOCKCHAIN_NAME.BSC,
     *   txTimestamp: 1658241570024,
     *   srxTxHash: '0xd2263ca82ac0fce606cb75df27d7f0dc94909d41a58c37563bd6772496cb8924'
     * };
     * const provider = CROSS_CHAIN_TRADE_TYPE.CELER;
     * const crossChainStatus = await sdk.crossChainStatusManager.getCrossChainStatus(tradeData, provider);
     * console.log('Source transaction status', crossChainStatus.srcTxStatus);
     * console.log('Destination transaction status', crossChainStatus.dstTxStatus);
     * ```
     * @param data Data needed to calculate statuses.
     * @param provider Cross-chain trade type.
     * @returns Object with transaction statuses.
     */
    CrossChainStatusManager.prototype.getCrossChainStatus = function (data, provider) {
        return __awaiter(this, void 0, void 0, function () {
            var crossChainStatus, fromBlockchain, srcTxHash, srcTxReceipt, srcTxStatus, dstTxStatus;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        crossChainStatus = {
                            srcTxStatus: cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN,
                            dstTxStatus: cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN
                        };
                        fromBlockchain = data.fromBlockchain, srcTxHash = data.srcTxHash;
                        return [4 /*yield*/, this.getTxReceipt(fromBlockchain, srcTxHash)];
                    case 1:
                        srcTxReceipt = _a.sent();
                        srcTxStatus = this.getSrcTxStatus(srcTxReceipt);
                        crossChainStatus.srcTxStatus = srcTxStatus;
                        return [4 /*yield*/, this.getDstTxStatus(srcTxStatus, srcTxReceipt, data, provider)];
                    case 2:
                        dstTxStatus = _a.sent();
                        if (dstTxStatus === cross_chain_tx_status_1.CrossChainTxStatus.FAIL && srcTxStatus === cross_chain_tx_status_1.CrossChainTxStatus.PENDING) {
                            crossChainStatus.srcTxStatus = cross_chain_tx_status_1.CrossChainTxStatus.FAIL;
                        }
                        crossChainStatus.dstTxStatus = dstTxStatus;
                        return [2 /*return*/, crossChainStatus];
                }
            });
        });
    };
    /**
     * Get cross-chain trade's source transaction status via receipt.
     * @param srcTxReceipt Transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getSrcTxStatus = function (srcTxReceipt) {
        if (srcTxReceipt === null) {
            return cross_chain_tx_status_1.CrossChainTxStatus.PENDING;
        }
        if (srcTxReceipt.status) {
            return cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS;
        }
        return cross_chain_tx_status_1.CrossChainTxStatus.FAIL;
    };
    /**
     * Get destination transaction status based on source transaction status,
     * source transaction receipt, trade data and provider type.
     * @param srcTxStatus Source transaction status.
     * @param srcTxReceipt Source transaction receipt.
     * @param tradeData Trade data.
     * @param provider Cross-chain trade type.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getDstTxStatus = function (srcTxStatus, srcTxReceipt, tradeData, provider) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (srcTxStatus === cross_chain_tx_status_1.CrossChainTxStatus.FAIL) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                        }
                        if (srcTxStatus === cross_chain_tx_status_1.CrossChainTxStatus.PENDING) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        return [4 /*yield*/, this.getDstTxStatusFnMap[provider].call(this, tradeData, srcTxReceipt)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    /**
     * Get Symbiosis trade dst transaction status.
     * @param data Trade data.
     * @param srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getSymbiosisDstSwapStatus = function (data, srcTxReceipt) {
        return __awaiter(this, void 0, void 0, function () {
            var symbiosisTxIndexingTimeSpent, srcChainId, _a, dstTxStatus, dstHash, error_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        symbiosisTxIndexingTimeSpent = Date.now() > data.txTimestamp + 30000;
                        if (!symbiosisTxIndexingTimeSpent) return [3 /*break*/, 4];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        srcChainId = core_1.BlockchainsInfo.getBlockchainByName(data.fromBlockchain).id;
                        return [4 /*yield*/, injector_1.Injector.httpClient.get("https://api.symbiosis.finance/crosschain/v1/tx/".concat(srcChainId, "/").concat(srcTxReceipt.transactionHash))];
                    case 2:
                        _a = _b.sent(), dstTxStatus = _a.status.text, dstHash = _a.tx.hash;
                        if (dstTxStatus === symbiosis_swap_status_1.SymbiosisSwapStatus.PENDING ||
                            dstTxStatus === symbiosis_swap_status_1.SymbiosisSwapStatus.NOT_FOUND) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        if (dstTxStatus === symbiosis_swap_status_1.SymbiosisSwapStatus.STUCKED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.REVERT];
                        }
                        if (dstTxStatus === symbiosis_swap_status_1.SymbiosisSwapStatus.REVERTED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FALLBACK];
                        }
                        if (dstTxStatus === symbiosis_swap_status_1.SymbiosisSwapStatus.SUCCESS) {
                            if (data.toBlockchain !== core_1.BLOCKCHAIN_NAME.BITCOIN) {
                                return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                            }
                            return [2 /*return*/, this.getBitcoinStatus(dstHash)];
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _b.sent();
                        console.debug('[Symbiosis Trade] Error retrieving dst tx status', error_1);
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 4: return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                }
            });
        });
    };
    /**
     * Get Li-fi trade dst transaction status.
     * @param data Trade data.
     * @param srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getLifiDstSwapStatus = function (data, srcTxReceipt) {
        return __awaiter(this, void 0, void 0, function () {
            var params, status, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!data.lifiBridgeType) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        params = {
                            bridge: data.lifiBridgeType,
                            fromChain: core_1.BlockchainsInfo.getBlockchainByName(data.fromBlockchain).id,
                            toChain: core_1.BlockchainsInfo.getBlockchainByName(data.toBlockchain).id,
                            txHash: srcTxReceipt.transactionHash
                        };
                        return [4 /*yield*/, injector_1.Injector.httpClient.get('https://li.quest/v1/status', { params: params })];
                    case 2:
                        status = (_a.sent()).status;
                        if (status === lifi_swap_status_1.LifiSwapStatus.DONE) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        if (status === lifi_swap_status_1.LifiSwapStatus.FAILED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                        }
                        if (status === lifi_swap_status_1.LifiSwapStatus.INVALID) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN];
                        }
                        if (status === lifi_swap_status_1.LifiSwapStatus.NOT_FOUND || status === lifi_swap_status_1.LifiSwapStatus.PENDING) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN];
                    case 3:
                        error_2 = _a.sent();
                        console.debug('[Li-fi Trade] error retrieving tx status', error_2);
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get Celer trade dst transaction status.
     * @param data Trade data.
     * @param srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getCelerDstSwapStatus = function (data, srcTxReceipt) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var requestLog, eightHours, dstTxStatus, _c, error_3;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 2, , 3]);
                        requestLog = decode_logs_1.LogsDecoder.decodeLogs(celer_cross_chain_event_statuses_abi_1.celerCrossChainEventStatusesAbi, srcTxReceipt)[0];
                        if (!requestLog) {
                            eightHours = 60 * 60 * 1000 * 8;
                            if (!requestLog && Date.now() > data.txTimestamp + eightHours) {
                                return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                            }
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        _c = Number;
                        return [4 /*yield*/, injector_1.Injector.web3PublicService
                                .getWeb3Public(data.toBlockchain)
                                .callContractMethod(celer_cross_chain_contracts_addresses_1.celerCrossChainContractsAddresses[data.toBlockchain], celer_cross_chain_contract_abi_1.celerCrossChainContractAbi, 'processedTransactions', {
                                methodArguments: [
                                    (_b = (_a = requestLog === null || requestLog === void 0 ? void 0 : requestLog.params) === null || _a === void 0 ? void 0 : _a.find(function (param) { return param.name === 'id'; })) === null || _b === void 0 ? void 0 : _b.value
                                ]
                            })];
                    case 1:
                        dstTxStatus = _c.apply(void 0, [_d.sent()]);
                        if (dstTxStatus === celer_swap_status_enum_1.CelerSwapStatus.NULL) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        if (dstTxStatus === celer_swap_status_enum_1.CelerSwapStatus.FAILED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                        }
                        if (dstTxStatus === celer_swap_status_enum_1.CelerSwapStatus.SUCCESS) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN];
                    case 2:
                        error_3 = _d.sent();
                        console.debug('[Celer Trade] error retrieving tx status', error_3);
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get Rubic trade dst transaction status.
     * @param data Trade data.
     * @param srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getRubicDstSwapStatus = function (data, srcTxReceipt) {
        return __awaiter(this, void 0, void 0, function () {
            var dstTxStatus, _a, error_4;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        _a = Number;
                        return [4 /*yield*/, injector_1.Injector.web3PublicService
                                .getWeb3Public(data.toBlockchain)
                                .callContractMethod(rubic_cross_chain_contracts_addresses_1.rubicCrossChainContractsAddresses[data.toBlockchain], processed_transactios_method_abi_1.PROCESSED_TRANSACTION_METHOD_ABI, 'processedTransactions', { methodArguments: [srcTxReceipt.transactionHash] })];
                    case 1:
                        dstTxStatus = _a.apply(void 0, [_b.sent()]);
                        if (dstTxStatus === rubic_swap_status_enum_1.RubicSwapStatus.NULL) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        if (dstTxStatus === rubic_swap_status_enum_1.RubicSwapStatus.PROCESSED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        if (dstTxStatus === rubic_swap_status_enum_1.RubicSwapStatus.REVERTED) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.UNKNOWN];
                    case 2:
                        error_4 = _b.sent();
                        console.debug('[Rubic Trade] Error retrieving tx status', error_4);
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get transaction receipt.
     * @param blockchain Blockchain name.
     * @param txHash Transaction hash.
     * @returns Transaction receipt.
     */
    CrossChainStatusManager.prototype.getTxReceipt = function (blockchain, txHash) {
        return __awaiter(this, void 0, void 0, function () {
            var receipt, error_5;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, injector_1.Injector.web3PublicService
                                .getWeb3Public(blockchain)
                                .getTransactionReceipt(txHash)];
                    case 1:
                        receipt = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_5 = _a.sent();
                        console.debug('Error retrieving src tx receipt', { error: error_5, txHash: txHash });
                        receipt = null;
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/, receipt];
                }
            });
        });
    };
    /**
     * Get DeBridge trade dst transaction status.
     * @param _data Trade data.
     * @param srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getDebridgeDstSwapStatus = function (_data, srcTxReceipt) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a, _b, send, _c, claim, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _e.trys.push([0, 2, , 3]);
                        params = { filter: srcTxReceipt.transactionHash, filterType: 1 };
                        return [4 /*yield*/, this.httpClient.get('https://api.debridge.finance/api/Transactions/GetFullSubmissionInfo', { params: params })];
                    case 1:
                        _a = _e.sent(), _b = _a.send, send = _b === void 0 ? null : _b, _c = _a.claim, claim = _c === void 0 ? null : _c;
                        if (!send || !claim) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                        }
                        if (claim === null || claim === void 0 ? void 0 : claim.transactionHash) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                    case 2:
                        _d = _e.sent();
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Get Via trade dst transaction status.
     * @param data Trade data.
     * @param _srcTxReceipt Source transaction receipt.
     * @returns Cross-chain transaction status.
     */
    CrossChainStatusManager.prototype.getViaDstSwapStatus = function (data, _srcTxReceipt) {
        return __awaiter(this, void 0, void 0, function () {
            var txStatusResponse, status, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, new router_sdk_1.Via(via_default_api_key_1.VIA_DEFAULT_CONFIG).checkTx({
                                actionUuid: data.viaUuid
                            })];
                    case 1:
                        txStatusResponse = _b.sent();
                        status = txStatusResponse.event;
                        if (status === via_swap_status_1.ViaSwapStatus.SUCCESS) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        if (status === via_swap_status_1.ViaSwapStatus.FAIL) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.FAIL];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 2:
                        _a = _b.sent();
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @internal
     * Get transaction status in bitcoin network;
     * @param hash Bitcoin transaction hash.
     */
    CrossChainStatusManager.prototype.getBitcoinStatus = function (hash) {
        return __awaiter(this, void 0, void 0, function () {
            var bitcoinTransactionStatus, btcStatusApi, _a, isCompleted;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        btcStatusApi = 'https://blockchain.info/rawtx/';
                        return [4 /*yield*/, this.httpClient.get("".concat(btcStatusApi).concat(hash))];
                    case 1:
                        bitcoinTransactionStatus = _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        _a = _b.sent();
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                    case 3:
                        isCompleted = (bitcoinTransactionStatus === null || bitcoinTransactionStatus === void 0 ? void 0 : bitcoinTransactionStatus.block_index) !== undefined;
                        if (isCompleted) {
                            return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.SUCCESS];
                        }
                        return [2 /*return*/, cross_chain_tx_status_1.CrossChainTxStatus.PENDING];
                }
            });
        });
    };
    return CrossChainStatusManager;
}());
exports.CrossChainStatusManager = CrossChainStatusManager;
