// rubic-sdk/lib/features/cross-chain/providers/common/celer-rubic/cross-chain-contract-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossChainContractData = void 0;
var injector_1 = require("../../../../../core/sdk/injector");
var common_1 = require("../../../../../common");
var CrossChainContractData = /** @class */ (function () {
    function CrossChainContractData(providersData, blockchain, address) {
        this.providersData = providersData;
        this.blockchain = blockchain;
        this.address = address;
        this.web3Public = injector_1.Injector.web3PublicService.getWeb3Public(blockchain);
    }
    CrossChainContractData.prototype.getProvider = function (providerIndex) {
        var _a, _b;
        var provider = (_b = (_a = this.providersData) === null || _a === void 0 ? void 0 : _a[providerIndex]) === null || _b === void 0 ? void 0 : _b.provider;
        if (!provider) {
            throw new common_1.RubicSdkError('Provider has to be defined');
        }
        return provider;
    };
    return CrossChainContractData;
}());
exports.CrossChainContractData = CrossChainContractData;
