// @renproject/chains-bitcoin/module/script/script.js
import { utils } from "@renproject/utils";
import { hash160 } from "../utils/utils";
import { Opcode } from "./opcodes";
const checksum = (hash) => utils.sha256(utils.sha256(hash)).slice(0, 4);
export class Script {
    constructor() {
        this.OP = Opcode;
        this.addOp = (op) => {
            this.script = new Uint8Array([...this.script, op]);
            return this;
        };
        this.addData = (data) => {
            this.script = new Uint8Array([...this.script, data.length, ...data]);
            return this;
        };
        this.bytes = () => {
            return this.script;
        };
        this.toScriptHashOut = () => {
            return new Script()
                .addOp(Script.OP.OP_HASH160)
                .addData(hash160(this.bytes()))
                .addOp(Script.OP.OP_EQUAL)
                .bytes();
        };
        this.toAddress = (prefix) => {
            // Hash
            const hash = hash160(this.bytes());
            // Prepend prefix
            const hashWithPrefix = utils.concat([prefix, hash]);
            // Append checksum
            const hashWithChecksum = utils.concat([
                hashWithPrefix,
                checksum(hashWithPrefix),
            ]);
            return hashWithChecksum;
        };
        this.script = new Uint8Array();
    }
}
Script.OP = Opcode;
