// @mui/joy/node/styles/CssVarsProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useColorScheme = exports.shouldSkipGeneratingVar = exports.getInitColorSchemeScript = exports.CssVarsProvider = void 0;

var _system = require("@mui/system");

var _extendTheme = _interopRequireDefault(require("./extendTheme"));

const shouldSkipGeneratingVar = keys => !!keys[0].match(/(typography|variants|focus|breakpoints)/);

exports.shouldSkipGeneratingVar = shouldSkipGeneratingVar;
const {
  CssVarsProvider,
  useColorScheme,
  getInitColorSchemeScript
} = (0, _system.unstable_createCssVarsProvider)({
  theme: (0, _extendTheme.default)(),
  attribute: 'data-joy-color-scheme',
  modeStorageKey: 'joy-mode',
  colorSchemeStorageKey: 'joy-color-scheme',
  defaultColorScheme: {
    light: 'light',
    dark: 'dark'
  },
  shouldSkipGeneratingVar
});
exports.getInitColorSchemeScript = getInitColorSchemeScript;
exports.useColorScheme = useColorScheme;
exports.CssVarsProvider = CssVarsProvider;
