// @renproject/chains-bitcoin/module/zcash.js
import { RenNetwork } from "@renproject/utils";
import { Blockbook } from "./APIs/blockbook";
import { Blockchair, BlockchairNetwork } from "./APIs/blockchair";
import { Insight } from "./APIs/insight";
import { SoChain, SoChainNetwork } from "./APIs/sochain";
import { BitcoinBaseChain } from "./base";
import { resolveBitcoinNetworkConfig, SoChainExplorer } from "./utils/utils";
const ZcashMainnet = {
    label: "Zcash",
    selector: "Zcash",
    nativeAsset: {
        name: "Zcash",
        symbol: "ZEC",
        decimals: 8,
    },
    averageConfirmationTime: 75,
    explorer: SoChainExplorer("zcash", "ZEC"),
    p2shPrefix: new Uint8Array([0x1c, 0xbd]),
    providers: [
        new Blockbook("https://zecblockexplorer.com/api/"),
        new Blockchair(BlockchairNetwork.ZCASH),
        { api: new SoChain(SoChainNetwork.ZEC), priority: 15 },
        { api: new Insight("https://explorer.z.cash/api/"), priority: 20 },
        { api: new Insight("https://zechain.net/api/v1/"), priority: 20 }, // TODO: test again, currently broken
    ],
    // validateAddress: (address: string) =>
    //     validateAddress(address, "ZEC", "mainnet"),
};
const ZcashTestnet = {
    label: "Zcash Testnet",
    selector: "Zcash",
    nativeAsset: {
        name: "Testnet Zcash",
        symbol: "ZEC",
        decimals: 8,
    },
    averageConfirmationTime: 75,
    isTestnet: true,
    explorer: SoChainExplorer("testnet/zcash", "ZECTEST"),
    p2shPrefix: new Uint8Array([0x1c, 0xba]),
    providers: [
        new Insight("https://explorer.testnet.z.cash/api/"),
        { api: new SoChain(SoChainNetwork.ZECTEST), priority: 15 },
    ],
    // validateAddress: (address: string) =>
    //     validateAddress(address, "ZEC", "testnet"),
};
export class Zcash extends BitcoinBaseChain {
    constructor({ network }) {
        super({
            network: resolveBitcoinNetworkConfig(Zcash.configMap, network),
        });
        this.configMap = Zcash.configMap;
        this.assets = Zcash.assets;
    }
}
Zcash.chain = "Zcash";
Zcash.configMap = {
    [RenNetwork.Mainnet]: ZcashMainnet,
    [RenNetwork.Testnet]: ZcashTestnet,
    [RenNetwork.Devnet]: ZcashTestnet,
};
Zcash.assets = {
    ZEC: "ZEC",
};
