// @renproject/utils/types/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxStatusIndex = exports.TxStatus = void 0;
var TxStatus;
(function (TxStatus) {
    // TxStatusNil is used for transactions that have not been seen, or are
    // otherwise unknown.
    TxStatus["TxStatusNil"] = "nil";
    // TxStatusConfirming is used for transactions that are currently waiting
    // for their underlying blockchain transactions to be confirmed.
    TxStatus["TxStatusConfirming"] = "confirming";
    // TxStatusPending is used for transactions that are waiting for consensus
    // to be reached on when the transaction should be executed.
    TxStatus["TxStatusPending"] = "pending";
    // TxStatusExecuting is used for transactions that are currently being
    // executed.
    TxStatus["TxStatusExecuting"] = "executing";
    // TxStatusReverted is used for transactions that were reverted during
    // execution.
    TxStatus["TxStatusReverted"] = "reverted";
    // TxStatusDone is used for transactions that have been successfully
    // executed.
    TxStatus["TxStatusDone"] = "done";
})(TxStatus = exports.TxStatus || (exports.TxStatus = {}));
exports.TxStatusIndex = {
    [TxStatus.TxStatusNil]: 0,
    [TxStatus.TxStatusConfirming]: 1,
    [TxStatus.TxStatusPending]: 2,
    [TxStatus.TxStatusExecuting]: 3,
    [TxStatus.TxStatusReverted]: 4,
    [TxStatus.TxStatusDone]: 5,
};
