// @renproject/chains-ethereum/module/polygon.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { RenNetwork } from "@renproject/utils";
import { EthereumBaseChain } from "./base";
import { resolveEVMNetworkConfig } from "./utils/generic";
const configMap = {
    [RenNetwork.Mainnet]: {
        selector: "Polygon",
        nativeAsset: { name: "Matic", symbol: "MATIC", decimals: 18 },
        averageConfirmationTime: 4,
        config: {
            chainId: "0x89",
            chainName: "Polygon Mainnet",
            nativeCurrency: { name: "MATIC", symbol: "MATIC", decimals: 18 },
            rpcUrls: [
                "https://polygon-rpc.com",
                "https://rpc-mainnet.matic.network",
                "https://matic-mainnet.chainstacklabs.com",
                "https://rpc-mainnet.maticvigil.com",
                "https://rpc-mainnet.matic.quiknode.pro",
                "https://matic-mainnet-full-rpc.bwarelabs.com",
                "https://matic-mainnet-archive-rpc.bwarelabs.com",
                "https://poly-rpc.gateway.pokt.network",
                "https://rpc.ankr.com/polygon",
                "https://polygon-mainnet.public.blastapi.io",
            ],
            blockExplorerUrls: ["https://polygonscan.com"],
        },
        logRequestLimit: 1000,
        addresses: {
            GatewayRegistry: "0xf36666C230Fa12333579b9Bd6196CB634D6BC506",
            BasicBridge: "0x82DF02A52E2e76C0c233367f2fE6c9cfe51578c5",
        },
    },
    [RenNetwork.Testnet]: {
        selector: "Polygon",
        isTestnet: true,
        nativeAsset: { name: "Testnet Matic", symbol: "MATIC", decimals: 18 },
        averageConfirmationTime: 4,
        config: {
            chainId: "0x13881",
            chainName: "Mumbai",
            nativeCurrency: { name: "MATIC", symbol: "MATIC", decimals: 18 },
            rpcUrls: [
                "https://matic-mumbai.chainstacklabs.com",
                "https://rpc-mumbai.maticvigil.com",
                "https://matic-testnet-archive-rpc.bwarelabs.com",
            ],
            blockExplorerUrls: ["https://mumbai.polygonscan.com"],
        },
        logRequestLimit: 1000,
        addresses: {
            GatewayRegistry: "0x5076a1F237531fa4dC8ad99bb68024aB6e1Ff701",
            BasicBridge: "0xcb6bD6B6c7D7415C0157e393Bb2B6Def7555d518",
        },
    },
};
export class Polygon extends EthereumBaseChain {
    constructor(_a) {
        var { network } = _a, params = __rest(_a, ["network"]);
        super(Object.assign(Object.assign({}, params), { network: resolveEVMNetworkConfig(configMap, network) }));
        this.configMap = configMap;
        this.assets = Polygon.assets;
    }
}
// Static members.
Polygon.chain = "Polygon";
Polygon.configMap = configMap;
Polygon.assets = {
    MATIC: "MATIC",
};
