// @renproject/chains-ethereum/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveEVMNetworkConfig = exports.isEVMNetworkConfig = exports.resolveRpcEndpoints = exports.EthereumBaseChain = exports.EVM_ERROR = exports.EVMParam = void 0;
__exportStar(require("./utils/types"), exports);
var evmParams_1 = require("./utils/payloads/evmParams");
Object.defineProperty(exports, "EVMParam", { enumerable: true, get: function () { return evmParams_1.EVMParam; } });
var errors_1 = require("./utils/errors");
Object.defineProperty(exports, "EVM_ERROR", { enumerable: true, get: function () { return errors_1.EVM_ERROR; } });
var base_1 = require("./base");
Object.defineProperty(exports, "EthereumBaseChain", { enumerable: true, get: function () { return base_1.EthereumBaseChain; } });
var generic_1 = require("./utils/generic");
Object.defineProperty(exports, "resolveRpcEndpoints", { enumerable: true, get: function () { return generic_1.resolveRpcEndpoints; } });
Object.defineProperty(exports, "isEVMNetworkConfig", { enumerable: true, get: function () { return generic_1.isEVMNetworkConfig; } });
Object.defineProperty(exports, "resolveEVMNetworkConfig", { enumerable: true, get: function () { return generic_1.resolveEVMNetworkConfig; } });
// Chains
__exportStar(require("./arbitrum"), exports);
__exportStar(require("./avalanche"), exports);
__exportStar(require("./bsc"), exports);
__exportStar(require("./catalog"), exports);
__exportStar(require("./ethereum"), exports);
__exportStar(require("./fantom"), exports);
__exportStar(require("./goerli"), exports);
__exportStar(require("./kava"), exports);
__exportStar(require("./moonbeam"), exports);
__exportStar(require("./optimism"), exports);
__exportStar(require("./polygon"), exports);
