// @renproject/chains-ethereum/moonbeam.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Moonbeam = void 0;
const utils_1 = require("@renproject/utils");
const base_1 = require("./base");
const generic_1 = require("./utils/generic");
const configMap = {
    [utils_1.RenNetwork.Mainnet]: {
        selector: "Moonbeam",
        nativeAsset: { name: "Glimmer", symbol: "GLMR", decimals: 18 },
        averageConfirmationTime: 12,
        logRequestLimit: 2000,
        config: {
            chainId: "0x504",
            chainName: "Moonbeam",
            nativeCurrency: { name: "Glimmer", symbol: "GLMR", decimals: 18 },
            rpcUrls: ["https://rpc.api.moonbeam.network"],
            blockExplorerUrls: ["https://moonbeam.moonscan.io"],
        },
        addresses: {
            // TODO: Fill out once available.
            GatewayRegistry: "",
            BasicBridge: "",
        },
    },
    [utils_1.RenNetwork.Testnet]: {
        selector: "Moonbeam",
        nativeAsset: { name: "Glimmer", symbol: "GLMR", decimals: 18 },
        averageConfirmationTime: 12,
        logRequestLimit: 2000,
        config: {
            chainId: "0x507",
            chainName: "Moonbase Alpha",
            nativeCurrency: { name: "Dev", symbol: "DEV", decimals: 18 },
            rpcUrls: [
                "https://rpc.api.moonbase.moonbeam.network",
                "https://rpc.testnet.moonbeam.network",
            ],
            blockExplorerUrls: ["https://moonbase.moonscan.io"],
        },
        addresses: {
            GatewayRegistry: "0x5076a1F237531fa4dC8ad99bb68024aB6e1Ff701",
            BasicBridge: "0xcb6bD6B6c7D7415C0157e393Bb2B6Def7555d518",
        },
    },
};
class Moonbeam extends base_1.EthereumBaseChain {
    constructor(_a) {
        var { network } = _a, params = __rest(_a, ["network"]);
        super(Object.assign(Object.assign({}, params), { network: (0, generic_1.resolveEVMNetworkConfig)(configMap, network) }));
        this.configMap = configMap;
        this.assets = Moonbeam.assets;
    }
}
exports.Moonbeam = Moonbeam;
// Static members.
Moonbeam.chain = "Moonbeam";
Moonbeam.configMap = configMap;
Moonbeam.assets = {
    GLMR: "GLMR",
};
