// rubic-sdk/lib/features/cross-chain/providers/rubic-trade-provider/rubic-cross-chain-contract-trade/rubic-it-cross-chain-contract-trade/rubic-cross-chain-instant-trade/cross-chain-oneinch-trade.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrossChainOneinchTrade = void 0;
var core_1 = require("../../../../../../../core");
var constants_1 = require("../../../../../../instant-trades/dexes/common/oneinch-common/constants");
var wrapped_native_1 = require("../../../../celer-trade-provider/constants/wrapped-native");
var empty_address_1 = require("../../../../../../../core/blockchain/constants/empty-address");
var common_1 = require("../../../../../../../common");
var CrossChainOneinchTrade = /** @class */ (function () {
    function CrossChainOneinchTrade(instantTrade) {
        this.instantTrade = instantTrade;
        this.defaultDeadline = 999999999999999;
    }
    CrossChainOneinchTrade.prototype.getFirstPath = function () {
        var _a;
        if (!((_a = this.instantTrade.path) === null || _a === void 0 ? void 0 : _a[0])) {
            throw new common_1.RubicSdkError('Instant trade path has to be defined');
        }
        return this.instantTrade.path[0].address;
    };
    CrossChainOneinchTrade.prototype.getSecondPath = function () {
        return this.instantTrade.wrappedPath.map(function (token) { return core_1.Web3Pure.addressToBytes32(token.address); });
    };
    CrossChainOneinchTrade.prototype.modifyArgumentsForProvider = function (methodArguments, walletAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.instantTrade.encode({ fromAddress: walletAddress })];
                    case 1:
                        data = (_a.sent()).data;
                        if (!(methodArguments === null || methodArguments === void 0 ? void 0 : methodArguments[0])) {
                            throw new common_1.RubicSdkError('Method arguments array must not be empty');
                        }
                        methodArguments[0].push(data);
                        return [2 /*return*/];
                }
            });
        });
    };
    CrossChainOneinchTrade.prototype.getCelerSourceObject = function () {
        if (!this.instantTrade.transactionData) {
            throw new common_1.RubicSdkError("Can't estimate 1inch trade");
        }
        var dex = this.instantTrade.contractAddress;
        var _a = this.instantTrade.path.map(function (token) { return token.address; }), tokenIn = _a[0], restPath = _a.slice(1);
        var isOneInchNative = constants_1.oneinchApiParams.nativeAddress === tokenIn || tokenIn === empty_address_1.EMPTY_ADDRESS;
        var fromBlockchain = this.instantTrade.from
            .blockchain;
        var firstToken = isOneInchNative ? wrapped_native_1.wrappedNative[fromBlockchain] : tokenIn;
        if (!firstToken) {
            throw new common_1.RubicSdkError('First token has to be defined');
        }
        var secondToken = restPath === null || restPath === void 0 ? void 0 : restPath.pop();
        if (!secondToken) {
            throw new common_1.RubicSdkError('Second token has to be defined');
        }
        var path = [firstToken, secondToken];
        var amountOutMinimum = this.instantTrade.toTokenAmountMin.stringWeiAmount;
        return { dex: dex, path: path, data: this.instantTrade.transactionData, amountOutMinimum: amountOutMinimum };
    };
    CrossChainOneinchTrade.prototype.getCelerDestinationObject = function () {
        throw Error('[RUBIC SDK] 1Inch is not supported as target provider yet.');
    };
    return CrossChainOneinchTrade;
}());
exports.CrossChainOneinchTrade = CrossChainOneinchTrade;
